File: ts/src/priceLevel.ts
import { Order } from "./order";

class OrderNode {
  order: Order;
  prev: OrderNode | null =null;
  next: OrderNode | null =null;

  constructor(order: Order) {
    this.order = order;
  }
}
export class PriceLevel {
  private orders: Order[] = [];
  private head: OrderNode | null = null;
  private tail: OrderNode | null = null;
  private nodesById = new Map<string, OrderNode>();
  private runningQuantity = 0;


  enqueue(order: Order): void {
    if (this.nodesById.has(order.id)) {
      throw new Error(`Duplicate order ID: ${order.id}`);
    }

    const node = new OrderNode(order);

    if (this.tail === null)
    {
      this.tail = node;
      this.head = node;
    }
    else {
      node.prev = this.tail;
      this.tail.next = node;
      this.tail = node;
    }
    this.nodesById.set(order.id, node);
    this.runningQuantity += order.quantity;
  }

  dequeueFront(): Order | undefined {
    if (this.head === null) return undefined;
    const node = this.head;
    this.unlink(node);
    return node.order;
  }
  cancelOrder(orderId: string): boolean{
    const node = this.nodesById.get(orderId);
    if (!node) return false;
    this.unlink(node);
    return true
  }
  totalQuantity(): number{
    return this.runningQuantity
  }
  isEmpty(): boolean {
    return this.head === null;
}
  private unlink(node: OrderNode): void{
    if (node.prev) {
      node.prev.next = node.next
    }
    else {
      this.head = node.next;
    }

    if (node.next) {
      node.next.prev = node.prev;
    }
    else {
      this.tail = node.prev;
    }

    this.nodesById.delete(node.order.id)
    this.runningQuantity-=node.order.quantity
  }
}

File: cpp/src/PriceLevel.cpp
#include "PriceLevel.h"

#include <stdexcept>
#include <utility>

namespace quotient {

Order& PriceLevel::enqueue(Order order) {
  if (orders_by_id_.contains(order.id)) {
    throw std::invalid_argument("duplicate order id: " + order.id);
  }
  total_quantity_ += order.quantity;
  orders_.push_back(std::move(order));
  auto iterator = std::prev(orders_.end());
  orders_by_id_.emplace(iterator->id, iterator);
  return *iterator;
}

Order* PriceLevel::front() {
  return orders_.empty() ? nullptr : &orders_.front();
}

bool PriceLevel::remove(const std::string& order_id) {
  const auto found = orders_by_id_.find(order_id);
  if (found == orders_by_id_.end()) return false;
  total_quantity_ -= found->second->quantity;
  orders_.erase(found->second);
  orders_by_id_.erase(found);
  return true;
}

bool PriceLevel::reduce(const std::string& order_id, std::uint64_t quantity) {
  const auto found = orders_by_id_.find(order_id);
  if (found == orders_by_id_.end()) return false;
  if (quantity == 0 || quantity >= found->second->quantity) {
    throw std::invalid_argument("reduction must leave a positive remainder");
  }
  found->second->quantity -= quantity;
  total_quantity_ -= quantity;
  return true;
}

bool PriceLevel::empty() const noexcept { return orders_.empty(); }

std::uint64_t PriceLevel::total_quantity() const noexcept {
  return total_quantity_;
}

}  // namespace quotient
